Potential Fix for #72

bamine opened the pull request:
This is a potential fix for #72, I added a few more tests (in GroupByTests) to check if we match Spark's behaviour, but I need to add more for every aggregate function. I wanted to check if this is the correct approach.
@imarios @OlivierBlanvillain @kanterov please review !

kanterov commented on dataset/src/main/scala/frameless/functions/AggregateFunctions.scala:
```diff
@@ -20,74 +20,75 @@ trait AggregateFunctions {
+    variance: CatalystVariance[Sel, Agg],
+    encoder1: TypedEncoder[Sel],
+    encoder2: TypedEncoder[Option[Agg]]
+  ): TypedAggregateAndColumn[T, Option[Agg], Sel] = {
```
As far as I understand there is a bit of confusion with parameter names. I would expect such signature because selecting variance of empty dataset is None. (`TypedAggregateAndColumn[T, Option[Out], Out] is TypedColumn[T, Option[Out]]`) While when we do group by it we always have value for groups (`TypedAggregateAndColumn[T, Option[Out], Out] is TypedAggregate[T, Out]`)



```scala
def variance[T, In, Out: TypedEncoder](column: TypedColumn[T, In])(
    variance: CatalystVariance[In, Out]
): TypedAggregateAndColumn[T, Option[Out], Out] = {
  import column.aencoder
  new TypedAggregateAndColumn[T, Out, Option[Out]](untyped.variance(column.untyped))
}
```

kanterov commented on dataset/src/main/scala/frameless/TypedColumn.scala:
```diff
@@ -209,11 +209,16 @@ sealed trait TypedAggregate[T, A] extends UntypedExpression[T] {
   val aencoder: TypedEncoder[A],
   uencoder: TypedEncoder[U]
-) extends TypedColumn[T, U](expr) with TypedAggregate[T, A] {
+) extends TypedColumn[T, A](expr) with TypedAggregate[T, A] {
```
For instance, when we do `select(sum(xxx))` and `groupBy(yyy).agg(sum(xxx))` where `xxx` is `Long`, in `groupBy` we get `Long`, and in `select` we get `Option[Long]`

bamine commented:
Changed stuff according to @kanterov 's explanation, but not sure this is still relevant with #75 (even though I also replaced `Ordering` -> `CatalystOrdered` and `Variance` -> `CatalystVariance`)
